### PR TITLE
Basic structure

### PR DESCRIPTION
**Pull request by iTitus**

### CODE REVIEW

**Comment by iTitus** [comment]
I still need to fix pylint :)

File: .github/workflows/python-app.yml [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+          pip install --upgrade -r requirements.txt
+      - name: Lint with flake8
+        run: |
+          flake8 . --count --show-source --statistics
+      - name: Lint with pylint
+        run: |
+          pylint src tests
+      - name: Test with pytest
+        run: |
+          pytest
```
Comment (Segelzwerg) [inline .github/workflows/python-app.yml:42]: I think the step used in Whist-Server are better.
```suggestion
    - name: Python Pytest with Coverage
      run: |
        python3 -m pytest --cov=server --cov-report=xml
    - name: Upload to codecoverage
      uses: codecov/codecov-action@v1
      with:
        file: ./coverage.xml
        fail_ci_if_error: true
        token: ${{ secrets.CODECOV_TOKEN }}
```
We should also use a separate workflow for pylint as on fail we would get any information about the tests.

**Review by Segelzwerg — CHANGES_REQUESTED** [review]
Can you also provide tests?

**Review by Segelzwerg — CHANGES_REQUESTED** [review]
We should use the rcfile here

File: .github/workflows/pylint.yml [hunk 2]
```diff
@@ -0,0 +1,34 @@
+name: Python application
+
+on:
+  push:
+    branches: [ main, develop ]
+  pull_request:
+    branches: [ main ]
+
+jobs:
+  build:
```
Comment (Segelzwerg) [inline .github/workflows/pylint.yml:10]:
```suggestion
  pylint:
```

File: .github/workflows/python-app.yml [hunk 3]
```diff
@@ -0,0 +1,48 @@
+name: Python application
+
+on:
+  push:
+    branches: [ main, develop ]
+  pull_request:
+    branches: [ main ]
+
+jobs:
+  build:
```
Comment (Segelzwerg) [inline .github/workflows/python-app.yml:10]:
```suggestion
  test:
```

File: .github/workflows/pylint.yml [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+      - name: Install dependencies
+        run: |
+          python -m pip install --upgrade pip
+          pip install --upgrade -r requirements.txt
+      - name: Lint with flake8
+        run: |
+          flake8 . --count --show-source --statistics
+      - name: Lint with pylint
+        run: |
+          pylint src tests
```
Comment (Segelzwerg) [inline .github/workflows/pylint.yml:34]:
```suggestion
          pylint src tests --rcfile=.pylintrc
```

Comment (iTitus) [reply .github/workflows/pylint.yml:34]: Not necessary

Comment (Segelzwerg) [reply .github/workflows/pylint.yml:34]: I didn't know it will load it by default.

File: src/whist/card.py [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+        return self.value if isinstance(self.value, str) else self.value[1]
+
+    def __str__(self) -> str:
+        return self.label
+
+
+@final
+@dataclass(frozen=True)
+class Card:
+    """A plying card"""
```
Comment (Segelzwerg) [inline src/whist/card.py:166]:
```suggestion
    """A playing card"""
```

File: src/whist/card.py [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+class _OrderedEnum(Enum):
+    def __new__(cls, *args) -> NoReturn:
+        if len(args) == 1:
+            value = args[0]
+        else:
+            value = args
+
+        obj = object.__new__(cls)
+        obj._value_ = value
+        obj.__ordinal = len(cls.__members__)  # pylint: disable=protected-access
```
Comment (Segelzwerg) [inline src/whist/card.py:17]:
```suggestion
        # pylint: disable=protected-access
        obj.__ordinal = len(cls.__members__) 
```

File: src/whist/card.py [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+        raise KeyError
+
+    @property
+    def symbol(self) -> str:
+        """
+        Get the suit symbol.
+
+        :return: suit symbol
+        """
+        return self.value[0]  # pylint: disable=unsubscriptable-object
```
Comment (Segelzwerg) [inline src/whist/card.py:93]:
```suggestion
        # pylint: disable=unsubscriptable-object
        return self.value[0]
```

File: src/whist/card.py [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+
+    @property
+    def label(self) -> str:
+        """
+        Get the suit label.
+
+        :return: suit label
+        """
+        return self.value[1]  # pylint: disable=unsubscriptable-object
```
Comment (Segelzwerg) [inline src/whist/card.py:102]:
```suggestion
        # pylint: disable=unsubscriptable-object
        return self.value[1]
```

File: src/whist/card.py [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+        raise KeyError
+
+    @property
+    def short_label(self) -> str:
+        """
+        Get the short version of the rank label.
+
+        :return: short version of rank label
+        """
+        return self.value if isinstance(self.value, str) else self.value[0]  # pylint: disable=unsubscriptable-object
```
Comment (Segelzwerg) [inline src/whist/card.py:150]:
```suggestion
        # pylint: disable=unsubscriptable-object
        return self.value if isinstance(self.value, str) else self.value[0]
```

File: src/whist/card.py [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+
+    @property
+    def label(self) -> str:
+        """
+        Get the rank label.
+
+        :return: rank label
+        """
+        return self.value if isinstance(self.value, str) else self.value[1]  # pylint: disable=unsubscriptable-object
```
Comment (Segelzwerg) [inline src/whist/card.py:159]:
```suggestion
        # pylint: disable=unsubscriptable-object
        return self.value if isinstance(self.value, str) else self.value[1]
```

File: src/whist/deck.py [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+
+    def __str__(self) -> str:
+        return str(self.__cards)
+
+    def __repr__(self) -> str:
+        return f'Deck(cards={self.__cards!r})'
+
+    def __eq__(self, other: Any) -> bool:
+        if self.__class__ is other.__class__:
+            return self.__cards == other.__cards  # pylint: disable=protected-access
```
Comment (Segelzwerg) [inline src/whist/deck.py:60]:
```suggestion
            # pylint: disable=protected-access
            return self.__cards == other.__cards
```

File: src/whist/deck.py [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+        return NotImplemented
+
+    def __ne__(self, other: Any) -> bool:
+        if self.__class__ is other.__class__:
+            return self.__cards != other.__cards  # pylint: disable=protected-access
```
Comment (Segelzwerg) [inline src/whist/deck.py:65]:
```suggestion
            # pylint: disable=protected-access
            return self.__cards != other.__cards
```

File: src/whist/card.py (hunk 6 above)
Comment (iTitus) [reply src/whist/card.py:17]: why?

Comment (Segelzwerg) [reply src/whist/card.py]: I was under the impression you have to but the comment above the line for which you want to disable it.

File: .github/workflows/python-app.yml [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+      - name: Install dependencies
+        run: |
+          python -m pip install --upgrade pip
+          pip install --upgrade -r requirements.txt
+      - name: Lint with flake8
+        run: |
+          flake8 . --count --select=E9,F63,F7,F82 --show-source --statistics
+      - name: Lint with pylint
+        run: |
+          pylint src tests
```
Comment (Segelzwerg) [inline .github/workflows/python-app.yml:39]:
```suggestion
```

File: tests/whist/test_card.py [hunk 14]
```diff
@@ -0,0 +1,16 @@
...
+
+def test_rank_order():
+    assert Rank.NUM_2 < Rank.NUM_3 < Rank.NUM_4 < Rank.NUM_5 < Rank.NUM_6 < Rank.NUM_7 \
+           < Rank.NUM_8 < Rank.NUM_9 < Rank.NUM_10 < Rank.J < Rank.Q < Rank.K < Rank.A
+
+
+def test_card_equality():
+    c1 = Card(Suit.CLUBS, Rank.NUM_2)
+    c2 = Card(Suit.CLUBS, Rank.NUM_2)
+    assert c1 == c2
```
Comment (Segelzwerg) [inline tests/whist/test_card.py:16]: I think it is better to use unittests as they give you more features in more complicated scenarios. And with starting using them now will keep everything consistent.

File: src/whist/card.py [hunk 15]
```diff
@@ -0,0 +1,196 @@
...
+
+    def __le__(self, other: Any) -> bool:
+        if self.__class__ is other.__class__:
+            return self.ordinal <= other.ordinal
+        return NotImplemented
+
+    def __lt__(self, other: Any) -> bool:
+        if self.__class__ is other.__class__:
+            return self.ordinal < other.ordinal
+        return NotImplemented
```
Comment (Segelzwerg) [inline src/whist/card.py:58]: Is it possible to reduce duplication here?

File: src/whist/deck.py [hunk 16]
```diff
@@ -0,0 +1,89 @@
...
+
+    def remove(self, card: Card) -> NoReturn:
+        """
+        Remove a card from this deck.
+
+        :param card: card to remove
+        """
+        self.__cards.remove(card)
+
+    def __contains__(self, item: Any) -> bool:
```
Comment (Segelzwerg) [inline src/whist/deck.py:43]: Should we allow `Any` to be here? We could use `Card` directly?

File: setup.py [hunk 17]
```diff
@@ -1,11 +1,27 @@
-from setuptools import setup
+from setuptools import setup, find_packages
+
+with open('README.md', 'r', encoding='utf-8') as f:
+    long_description = f.read()
 
 setup(
-    name='Whist-Core',
+    name='whist-core',
     version='0.0.1',
```
Comment (Segelzwerg) [inline setup.py:8]:
```suggestion
    version='0.0.2',
```

there was already a tag uploaded to PyPI with `0.0.1` which was deleted by mistake, therefore the next release should be `0.0.2`

**Pull request merged**